app/scripts/modules/core/delivery/service/execution.service.ts

```diff
@@ -0,0 +1,358 @@
+        let match = (application.executions.data || []).find((test: IExecution) => test.id === execution.id);
+        if (!match || !match.stringVal || match.stringVal !== stringVal) {
+          execution.stringVal = stringVal;
+          this.executionsTransformer.transformExecution(application, execution);
+        }
+      });
+    }
+
+    private cleanExecutionForDiffing(execution: IExecution): void {
+      (execution.stages || []).forEach((stage: IExecutionStage) => this.removeInstances(stage));
+      if (execution.trigger && execution.trigger.parentExecution) {
+        (execution.trigger.parentExecution.stages || []).forEach((stage: IExecutionStage) => this.removeInstances(stage));
+      }
+    }
+
+    // these fields are never displayed in the UI, so don't retain references to them, as they consume a lot of memory
+    // on very large deployments
+    private removeInstances(stage: IExecutionStage): void {
+      if (stage.context) {
+        delete stage.context.instances;
+        delete stage.context.asg;
+        if (stage.context.targetReferences) {
+          stage.context.targetReferences.forEach((targetReference: any) => {
+            delete targetReference.instances;
+            delete targetReference.asg;
+          });
+        }
+      }
+    }
+
+    // remove these fields - they are not of interest when determining if the pipeline has changed
+    private jsonReplacer(key: string, value: any): any {
+      if (key === 'instances' || key === 'asg' || key === 'commits' || key === 'history' || key === '$$hashKey') {
+        return undefined;
+      }
+      return value;
+    }
+
+    public waitUntilNewTriggeredPipelineAppears(application: Application, pipelineName: string, triggeredPipelineId: string): IPromise<any> {
+      return this.getRunningExecutions(application.name).then((executions: IExecution[]) => {
+        const match = executions.find((execution) => execution.id === triggeredPipelineId);
+        const deferred = this.$q.defer();
+        if (match) {
+          application.executions.refresh().then(deferred.resolve);
+          return deferred.promise;
+        } else {
+          return this.$timeout(function() {
+            return this.waitUntilNewTriggeredPipelineAppears(application, pipelineName, triggeredPipelineId);
+          }, 1000);
+        }
+      });
+    }
+
+    private waitUntilPipelineIsCancelled(application: Application, executionId: string): IPromise<any> {
+      return this.waitUntilExecutionMatches(executionId, (execution: IExecution) => execution.status === 'CANCELED')
+        .then(() => application.executions.refresh());
+    }
+
+    private waitUntilPipelineIsDeleted(application: Application, executionId: string): IPromise<any> {
+      const deferred = this.$q.defer();
+      this.getExecution(executionId).then(
+        () => this.$timeout(() => this.waitUntilPipelineIsDeleted(application, executionId).then(deferred.resolve), 1000),
+        deferred.resolve
```

Comment:
any weird `this` binding behavior here? i.e. does it need to be `() => deferred.resolve()`?